### PR TITLE
chore(query): round the results for decimal division

### LINKED ISSUE #13825: wizard test: decimal precision issue
**Summary**

The precision issue is compared to snowflake, here only give how Databend to reproduce.
**Note: all the result is output by bendsql and snowsql.**
bendsql --query="<query>" 
snowsql --query "<query>" -o output_format=tsv -o header=false -o timing=false -o friendly=false

## Setup

https://github.com/datafuselabs/wizard/blob/main/checksb/sql/selects/bend/setup.sql

```
DROP TABLE IF EXISTS customers;
CREATE TABLE customers (
    customer_id INT UNSIGNED NOT NULL,
    customer_name VARCHAR NOT NULL,
    segment VARCHAR NOT NULL,
    create_timestamp DATE NOT NULL,
    active BOOLEAN NOT NULL
);

DROP TABLE IF EXISTS date_dim;
CREATE TABLE date_dim (
    date_key DATE NOT NULL,
    day_of_week TINYINT UNSIGNED NOT NULL,
    month TINYINT UNSIGNED NOT NULL,
    year SMALLINT UNSIGNED NOT NULL
);

DROP TABLE IF EXISTS products;
CREATE TABLE products
(
    product_id   INT UNSIGNED NOT NULL,
    product_name VARCHAR        NOT NULL,
    price        DECIMAL(10, 2) NOT NULL,
    category     VARCHAR        NOT NULL
);

DROP TABLE IF EXISTS sales;
CREATE TABLE sales (
    sale_id INT UNSIGNED NOT NULL,
    product_id INT UNSIGNED NOT NULL,
    customer_id INT UNSIGNED NOT NULL,
    sale_date DATE NOT NULL,
    quantity INT NOT NULL,
    net_paid DECIMAL(10, 2) NOT NULL
);

CREATE STAGE IF NOT EXISTS wizardbend
    URL = 's3://wizardbend/'
    CONNECTION = (ALLOW_ANONYMOUS = 'true');

COPY INTO customers
    FROM @wizardbend/selects/customers.parquet
    FILE_FORMAT = (TYPE = parquet);

COPY INTO date_dim
    FROM @wizardbend/selects/date_dim.parquet
    FILE_FORMAT = (TYPE = parquet);

COPY INTO products
    FROM @wizardbend/selects/products.parquet
    FILE_FORMAT = (TYPE = parquet);


COPY INTO sales
    FROM @wizardbend/selects/sales.parquet
    FILE_FORMAT = (TYPE = parquet);
```

## Queries

```
-- Query 3: Average sale amount per customer segment
SELECT c.segment, AVG(s.net_paid) AS avg_sale_amount
FROM sales s
         JOIN customers c ON s.customer_id = c.customer_id
GROUP BY c.segment
ORDER BY avg_sale_amount DESC
```
image

```
-- Query 7: Average product price by category
SELECT p.category, AVG(p.price) AS avg_price
FROM products p
GROUP BY p.category
ORDER BY avg_price DESC
```
image

```
-- Query 13: Total sales and average quantity sold per product
SELECT p.product_id, p.product_name, SUM(s.net_paid) AS total_sales, AVG(s.quantity) AS avg_quantity_sold
FROM sales s
         JOIN products p ON s.product_id = p.product_id
GROUP BY p.product_id, p.product_name
ORDER BY total_sales DESC
    LIMIT 10
```
image

```
-- Query 21: Products with sales above average in their category
WITH CategoryAverage AS (
    SELECT
        p.category,
        AVG(s.net_paid) AS avg_sales
    FROM
        sales s
            JOIN
        products p ON s.product_id = p.product_id
    GROUP BY
        p.category
)
SELECT
    p.product_id,
    p.product_name,
    p.category,
    SUM(s.net_paid) AS total_sales,
    ca.avg_sales
FROM
    sales s
        JOIN
    products p ON s.product_id = p.product_id
        JOIN
    CategoryAverage ca ON p.category = ca.category
GROUP BY
    p.product_id, p.product_name, p.category, ca.avg_sales
HAVING
        SUM(s.net_paid) > ca.avg_sales
ORDER BY
    p.product_id, total_sales DESC LIMIT 10
```
image

## Check scripts

The full check queries: https://github.com/datafuselabs/wizard/blob/main/checksb/sql/selects/check.sql

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

_Briefly describe what this PR aims to solve. Include background context that will help reviewers understand the purpose of the PR._

Fixes #13825

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14109)

### CODE REVIEW

File: src/query/functions/src/scalars/decimal/arithmetic.rs [hunk 1]
```diff
@@ -85,22 +85,23 @@ macro_rules! binary_decimal {
             let scale_a = $left.scale();
             let scale_b = $right.scale();
 
-            let (scale_mul, scale_div) = if scale_b + $size.scale > scale_a {
-                (scale_b + $size.scale - scale_a, 0)
-            } else {
-                (0, scale_b + $size.scale - scale_a)
-            };
 
+            // Note: the result scale is larger the left's scale
```
Comment (RinChanNOWWW) [inline src/query/functions/src/scalars/decimal/arithmetic.rs:89]: Typo: larger than.

File: src/query/functions/src/scalars/decimal/arithmetic.rs [hunk 2]
```diff
@@ -85,22 +85,23 @@ macro_rules! binary_decimal {
...
+                } else if a.is_negative() == b.is_negative() {
+                    result.push((a * multiplier + b / 2).div(b));
                 } else {
-                    result.push((a * multiplier).div(b) / div);
+                    result.push((a * multiplier - b / 2).div(b));
                 }
+                // 5 2 --> 3
+                // 5, -2 ---> -3
+                // -5, 2 --> -3
+                // -5, -2 --> 3
```
Comment (RinChanNOWWW) [inline src/query/functions/src/scalars/decimal/arithmetic.rs:104]: What does this mean? The comment is confusing.

File: src/query/functions/src/scalars/decimal/arithmetic.rs [hunk 3]
```diff
@@ -85,21 +85,24 @@ macro_rules! binary_decimal {
             let scale_a = $left.scale();
             let scale_b = $right.scale();
 
-            let (scale_mul, scale_div) = if scale_b + $size.scale > scale_a {
-                (scale_b + $size.scale - scale_a, 0)
-            } else {
-                (0, scale_b + $size.scale - scale_a)
-            };
 
+            // Note: the result scale is always larger than the left scale
```
Comment (RinChanNOWWW) [inline src/query/functions/src/scalars/decimal/arithmetic.rs:89]: How about adding a debug assertion?

**Review by RinChanNOWWW — APPROVED** [review]

**Review by b41sh — APPROVED** [review]

File: src/query/functions/src/scalars/decimal/arithmetic.rs (hunk 3 above)
Comment (sundy-li) [reply src/query/functions/src/scalars/decimal/arithmetic.rs:89]: Ok，  Next PR.

**Comment by BohuTANG** [comment]
wizard passed 👍

**Pull request merged**